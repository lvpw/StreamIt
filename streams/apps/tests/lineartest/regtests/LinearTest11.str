/** Test combining pipelines when they don't rate match**/
void->void pipeline LinearTest11 {
  add FloatSource();
  add LinearPipe();
  add LinearPipe2();
  add FloatSink();
}

void->float filter FloatSource {
  float x;
  init {
    x = 0;
  }
  work push 1 {
    push(x);
    x = x + 1;
  }
}

float->void filter FloatSink {
  work pop 1 {
    println(pop());
  }
}

/** Pipeline of linear filters whose rates don't match. **/
float->float pipeline LinearPipe {
  add ParamIdentity(1);
  add LinearFilter1(4,4);
  add ParamIdentity(2);
}

/** pipeline with some filters that don't just do identity multiplications. **/
float->float pipeline LinearPipe2 {
  add LinearFilter1(3,2);
  add LinearFilter2();
}


/** 
 * Simple filter that just computes a small matrix. 
 * Still keeping peek==pop.
 **/
float->float filter LinearFilter1(int numPeeks, int numPushes) {
  work peek numPeeks pop numPeeks push numPushes{
    for (int i=0; i<numPushes; i++) {
      int sum = 0;
      for (int j=0; j<numPeeks; j++) {
	sum += peek(j)*((i+1)*(j + 1) +i);
      }
      push(sum);
    }
    for (int i=0; i<numPeeks; i++) {
      pop();
    }
  }
}

/** 
 * Another filter so that we can be sure that the matrix multiply is actually occuring
 * (because we are multiplying by the identity matrix earlier in the pipeline.
 **/
float->float filter LinearFilter2 {
  // calculates [[1 2][3 4][5 6]]
  work push 2 pop 3 {
    float t1 = pop();
    float t2 = pop();
    float t3 = pop();
    push(2*t3 + 4*t2 + 6*t1);
    push(1*t3 + 3*t2 + 5*t1);
  }
}













/** Parameterized identity filter to test resizing filters in the pipeline. **/
float->float filter ParamIdentity(int size) {
  work peek size pop size push size {
    for (int i=0; i<size; i++) {
      push(peek(i));
    }
    for (int i=0; i<size; i++) {
      pop();
    }
  }
}


