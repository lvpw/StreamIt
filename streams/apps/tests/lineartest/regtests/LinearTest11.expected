LinearFilter1_1.matrix
[[19.0+0.0i 14.0+0.0i 9.0+0.0i 4.0+0.0i]
 [15.0+0.0i 11.0+0.0i 7.0+0.0i 3.0+0.0i]
 [11.0+0.0i 8.0+0.0i 5.0+0.0i 2.0+0.0i]
 [7.0+0.0i 5.0+0.0i 3.0+0.0i 1.0+0.0i]]
LinearFilter1_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
LinearFilter1_2.matrix
[[7.0+0.0i 3.0+0.0i]
 [5.0+0.0i 2.0+0.0i]
 [3.0+0.0i 1.0+0.0i]]
LinearFilter1_2.vector
[[0.0+0.0i 0.0+0.0i]]
LinearFilter2_1.matrix
[[1.0+0.0i 2.0+0.0i]
 [3.0+0.0i 4.0+0.0i]
 [5.0+0.0i 6.0+0.0i]]
LinearFilter2_1.vector
[[0.0+0.0i 0.0+0.0i]]
LinearPipe2_1.matrix
[[16.0+0.0i 26.0+0.0i 0.0+0.0i 0.0+0.0i]
 [11.0+0.0i 18.0+0.0i 0.0+0.0i 0.0+0.0i]
 [6.0+0.0i 10.0+0.0i 0.0+0.0i 0.0+0.0i]
 [35.0+0.0i 42.0+0.0i 3.0+0.0i 6.0+0.0i]
 [25.0+0.0i 30.0+0.0i 2.0+0.0i 4.0+0.0i]
 [15.0+0.0i 18.0+0.0i 1.0+0.0i 2.0+0.0i]
 [0.0+0.0i 0.0+0.0i 36.0+0.0i 46.0+0.0i]
 [0.0+0.0i 0.0+0.0i 25.0+0.0i 32.0+0.0i]
 [0.0+0.0i 0.0+0.0i 14.0+0.0i 18.0+0.0i]]
LinearPipe2_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
LinearPipe_1.matrix
[[19.0+0.0i 14.0+0.0i 9.0+0.0i 4.0+0.0i]
 [15.0+0.0i 11.0+0.0i 7.0+0.0i 3.0+0.0i]
 [11.0+0.0i 8.0+0.0i 5.0+0.0i 2.0+0.0i]
 [7.0+0.0i 5.0+0.0i 3.0+0.0i 1.0+0.0i]]
LinearPipe_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
LinearWrapper_1.matrix
[[652.0+0.0i 1004.0+0.0i 12.0+0.0i 24.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [508.0+0.0i 784.0+0.0i 9.0+0.0i 18.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [364.0+0.0i 564.0+0.0i 6.0+0.0i 12.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [220.0+0.0i 344.0+0.0i 3.0+0.0i 6.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [685.0+0.0i 822.0+0.0i 476.0+0.0i 646.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [540.0+0.0i 648.0+0.0i 368.0+0.0i 500.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [395.0+0.0i 474.0+0.0i 260.0+0.0i 354.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [250.0+0.0i 300.0+0.0i 152.0+0.0i 208.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 266.0+0.0i 342.0+0.0i 347.0+0.0i 566.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 210.0+0.0i 270.0+0.0i 271.0+0.0i 442.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 154.0+0.0i 198.0+0.0i 195.0+0.0i 318.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 98.0+0.0i 126.0+0.0i 119.0+0.0i 194.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 1150.0+0.0i 1380.0+0.0i 238.0+0.0i 372.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 905.0+0.0i 1086.0+0.0i 182.0+0.0i 286.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 660.0+0.0i 792.0+0.0i 126.0+0.0i 200.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 415.0+0.0i 498.0+0.0i 70.0+0.0i 114.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 671.0+0.0i 860.0+0.0i 188.0+0.0i 306.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 529.0+0.0i 678.0+0.0i 145.0+0.0i 236.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 387.0+0.0i 496.0+0.0i 102.0+0.0i 166.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 245.0+0.0i 314.0+0.0i 59.0+0.0i 96.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 889.0+0.0i 1120.0+0.0i 64.0+0.0i 128.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 695.0+0.0i 876.0+0.0i 50.0+0.0i 100.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 501.0+0.0i 632.0+0.0i 36.0+0.0i 72.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 307.0+0.0i 388.0+0.0i 22.0+0.0i 44.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 1160.0+0.0i 1484.0+0.0i 64.0+0.0i 104.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 913.0+0.0i 1168.0+0.0i 48.0+0.0i 78.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 666.0+0.0i 852.0+0.0i 32.0+0.0i 52.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 419.0+0.0i 536.0+0.0i 16.0+0.0i 26.0+0.0i 0.0+0.0i 0.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 708.0+0.0i 980.0+0.0i 35.0+0.0i 70.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 551.0+0.0i 764.0+0.0i 27.0+0.0i 54.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 394.0+0.0i 548.0+0.0i 19.0+0.0i 38.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 237.0+0.0i 332.0+0.0i 11.0+0.0i 22.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 285.0+0.0i 342.0+0.0i 804.0+0.0i 1042.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 225.0+0.0i 270.0+0.0i 628.0+0.0i 814.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 165.0+0.0i 198.0+0.0i 452.0+0.0i 586.0+0.0i]
 [0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 105.0+0.0i 126.0+0.0i 276.0+0.0i 358.0+0.0i]]
LinearWrapper_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
ParamIdentity_1.matrix
[[1.0+0.0i]]
ParamIdentity_1.vector
[[0.0+0.0i]]
ParamIdentity_2.matrix
[[1.0+0.0i 0.0+0.0i]
 [0.0+0.0i 1.0+0.0i]]
ParamIdentity_2.vector
[[0.0+0.0i 0.0+0.0i]]
